Document StatefulShortestPath(All) and StatefulShortestPath(Into)

WilcoNeo opened the pull request:
This PR adds the following:
- Operator description of `StatefulShortestPath(All)` and `StatefulShortestPath(Into)`
- Planner decisions regarding `StatefulShortestPath(All)` and `StatefulShortestPath(Into)`

WilcoNeo committed:
Document StatefulShortestPath(All) and StatefulShortestPath(Into)

WilcoNeo commented:
I will ping the cypher-planner-team first for a review (I can't add any labels)

WilcoNeo commented on modules/ROOT/pages/planning-and-tuning/operators/operators-detail.adoc:
```diff
@@ -5920,6 +5920,132 @@ Total database accesses: 5, total allocated memory: 1488
+
+Runtime PIPELINED
+
+Runtime version DEV
```
We probably want to update `DEV` to an actual version and re-run the query on that version. Which one should I use?

sherfert replied:
Looking at the other plan outputs, they use `{neo4j-version-minor}`

WilcoNeo replied:
Is there a way that I can specify that version when running a test (for example in MiscAcceptanceTest.scala)?

WilcoNeo commented:
- Operator `StatefulShortestPath(Into)`
<img width="567" alt="Screenshot 2024-05-10 at 15 25 08" src="https://github.com/JPryce-Aklundh/docs-cypher/assets/162103508/8b7b62b4-3a6e-4040-8476-78ce9c0fa4bf">

- Operator `StatefulShortestPath(All)`
<img width="585" alt="Screenshot 2024-05-10 at 15 23 39" src="https://github.com/JPryce-Aklundh/docs-cypher/assets/162103508/32160496-26ad-4fae-859b-7f3edcc0e4be">

- Planning decisions
<img width="459" alt="Screenshot 2024-05-10 at 16 40 02" src="https://github.com/JPryce-Aklundh/docs-cypher/assets/162103508/62c82b16-ac3b-48d2-84e8-3d553bc9e51f">


<img width="565" alt="Screenshot 2024-05-10 at 16 33 25" src="https://github.com/JPryce-Aklundh/docs-cypher/assets/162103508/782013b0-3705-4e6b-8268-dc5f895ca13a">

WilcoNeo committed:
Document workaround to force using StatefulShortestPath(Into)

sherfert commented on modules/ROOT/pages/planning-and-tuning/operators/operators-detail.adoc:
```diff
@@ -5920,6 +5920,132 @@ Total database accesses: 5, total allocated memory: 1488
+The `StatefulShortestPath(Into)` operator finds shortest paths between two previously matched node variables.
+It uses a bidirectional breath first search (BFS) algorithm, which performs two BFS invocations at the same time, one from the left boundary node and one from the right boundary node.
+Once a node is found by both BFS invocations, which indicates that it can be reached from both boundary nodes, the algorithm successfully terminates.
+No valid path exists between both boundary nodes when both BFS invocations terminate without any overlap in the nodes that both invocations found.
```
```suggestion
When both BFS invocations terminate without any overlap in the nodes found, then there exists no valid path between both boundary nodes.
```

sherfert commented on modules/ROOT/pages/planning-and-tuning/operators/operators-detail.adoc:
```diff
@@ -5920,6 +5920,132 @@ Total database accesses: 5, total allocated memory: 1488
+MATCH
+  (chris:Person {name: 'Chris'}),
+  (stefan:Person {name: 'Stefan'})
+CALL {
+  WITH chris, stefan
+  MATCH
+    p = ALL SHORTEST (chris)(()-[]-()-[]-()){1,}(stefan)
+  RETURN p
+}
+RETURN p
```
I would suggest changing the `CREATE RANGE INDEX range_person_name FOR (p:Person) ON (p.name)` to a uniqueness constraint instead.

Like that, you hopefully will not need the `CALL`-workaround in this example.

sherfert commented on modules/ROOT/pages/planning-and-tuning/operators/operators-detail.adoc:
```diff
@@ -5920,6 +5920,132 @@ Total database accesses: 5, total allocated memory: 1488
+[source, cypher]
+----
+PROFILE
+MATCH (chris {name: 'Chris'})
```
You probably want a `Person` label here?

WilcoNeo replied:
Yes, good catch. This, in combination with the index on `Person.name`, solved my issue in https://trello.com/c/eQvCFdzF/2587-optimization-opportunity-for-statefulshortestpath

sherfert commented on modules/ROOT/pages/planning-and-tuning/operators/operators-detail.adoc:
```diff
@@ -5920,6 +5920,132 @@ Total database accesses: 5, total allocated memory: 1488
+----
+PROFILE
+MATCH (chris {name: 'Chris'})
+CALL {
+  WITH chris
+  MATCH
+    p = ALL SHORTEST (chris)(()-[]-()-[]-()){1,}(location:Location)
+  RETURN length(p) AS pathLength, location.name AS locationName
+}
```
Why the subquery?

WilcoNeo replied:
I thought that it would make it more clear by explicitly saying that the query should start by matching `chris`. Also to make it more in line with the example for `INTO`. 
However, the example for `INTO` now doesn't need a `CALL` subquery due to the new `UNIQUE` index. Therefore we can remove the `CALL` subqueries in both cases now

sherfert commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,84 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+Using `StatefulShortestPath(Into)` requires 30 bidirectional breath first search (BFS) invocations.
+Using `StatefulShortestPath(All)` requires only 3 unidirectional BFS invocations expanding from the right boundary node.
+
+In case you want the `StatefulShortestPath(Into)` to be chosen, then you can rewrite your query using a `CALL subquery`.
```
```suggestion
Sometimes this heuristic decision to favour `StatefulShortestPath(All)` can lead to worse query performance.
In case you want the `StatefulShortestPath(Into)` to be chosen, then you can rewrite your query using a `CALL` subquery.
```

sherfert commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,84 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+
+
+
+Rewriting the query using a `CALL subquery`
```
```suggestion
Rewriting the query using a `CALL` subquery
```

sherfert commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,84 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+RETURN p
+----
+
+makes use of `StatefulShortestPath(Into)`, because the CALL subroutine is executed for each incoming row.  Each incoming row matches `a` to exactly one station node and `b` to exactly one station node.
```
```suggestion
makes use of `StatefulShortestPath(Into)`, because the `CALL` subquery is executed for each incoming row.  Each incoming row matches `a` to exactly one station node and `b` to exactly one station node.
```

sherfert commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,84 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+
+[[planning-StatefulShortestPath-operator]]
+== Planning the StatefulShortestPath-operator
+See xref:planning-and-tuning/operators/operators-detail.adoc#query-plan-stateful-shortest-path-into[the operator-detail page] for a description of the `StatefulShortestPath(All)` and `StatefulShortestPath(Into)` operators.
```
I would still add one sentence about the algorithms here and then point to the link for further details:

```
The `StatefulShortestPath(All)` uses a unidirectional breadth first search algorithm, while the `StatefulShortestPath(Into)` uses a bidirectional breadth first search algorithm.
```

sherfert commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,84 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+== Planning the StatefulShortestPath-operator
+See xref:planning-and-tuning/operators/operators-detail.adoc#query-plan-stateful-shortest-path-into[the operator-detail page] for a description of the `StatefulShortestPath(All)` and `StatefulShortestPath(Into)` operators.
+
+When the subexpression of the statefulShortestPath-operator matches only one of both boundary nodes, then a `StatefulShortestPath(All)` is planned.
+
+When the subexpression of the statefulShortestPath-operator matches both boundary nodes, then a `StatefulShortestPath(Into)` is planned only when both boundary nodes match to at most one data node. Otherwise, this plan is avoided when possible.
+Another plan will be favoured where the subexpression matches only one of both boundary nodes followed by an `StatefulShortestPath(All)`.
```
I find these paragraphs confusing. 

```
When the subexpression of the statefulShortestPath-operator matches only one of both boundary nodes
```

When would that happen?
What is "the subexpression of the statefulShortestPath-operator"?

---

I would phrase it along these lines:

```
The planner decides between the two variants using this heuristic:

* When both boundary nodes having been matched in a `MATCH` clause prior to the `MATCH` clause containing the shortest paths query, `StatefulShortestPath(Into)` is planned.
* Otherwise, if both boundary nodes match to at most one data node, `StatefulShortestPath(Into)` is planned.
* Otherwise, `StatefulShortestPath(All)` is planned.

Sometimes the planner cannot make reliable estimations about how many data nodes a patern nodes will match. Consider using a uniqueness constraint <insert link there> where applicable to help the planner getting more reliable estimates.
```

WilcoNeo committed:
Apply feedback
- Add unique constraint and Person label
- Remove CALL subqueries in operator-detail example queries
- Improve description of heuristic decision between ALL and INTO

sherfert commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -542,18 +542,24 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
 
-When the subexpression of the statefulShortestPath-operator matches both boundary nodes, then a `StatefulShortestPath(Into)` is planned only when both boundary nodes match to at most one data node. Otherwise, this plan is avoided when possible.
-Another plan will be favoured where the subexpression matches only one of both boundary nodes followed by an `StatefulShortestPath(All)`.
+* When both boundary nodes have been matched in `MATCH` clause prior to the `MATCH` clause containing the shortest paths query, `StatefulShortestPath(Into)` is planned
```
```suggestion
* When both boundary nodes have been matched in a `MATCH` clause prior to the `MATCH` clause containing the shortest paths query, `StatefulShortestPath(Into)` is planned.
```

sherfert commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -542,18 +542,24 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
-When the subexpression of the statefulShortestPath-operator matches both boundary nodes, then a `StatefulShortestPath(Into)` is planned only when both boundary nodes match to at most one data node. Otherwise, this plan is avoided when possible.
-Another plan will be favoured where the subexpression matches only one of both boundary nodes followed by an `StatefulShortestPath(All)`.
+* When both boundary nodes have been matched in `MATCH` clause prior to the `MATCH` clause containing the shortest paths query, `StatefulShortestPath(Into)` is planned
+* Otherwise, if both boundary nodes match to at most one data node, then `StatefulShortestPath(Into)` is planned
+* Otherwise, `StatefulShortestPath(All)` is planned
```
```suggestion
* Otherwise, if both boundary nodes match to at most one data node, then `StatefulShortestPath(Into)` is planned.
* Otherwise, `StatefulShortestPath(All)` is planned.
```

sherfert commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -542,18 +542,24 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+* Otherwise, if both boundary nodes match to at most one data node, then `StatefulShortestPath(Into)` is planned
+* Otherwise, `StatefulShortestPath(All)` is planned
+
+Sometimes the planner cannot make reliable estimations about how many data nodes a pattern nodes will match.
```
```suggestion
Sometimes the planner cannot make reliable estimations about how many data nodes a pattern node will match.
```

sherfert reviewed (approved):
Only minor cosmetic suggestions left. :)

WilcoNeo committed:
Update modules/ROOT/pages/patterns/shortest-paths.adoc

Co-authored-by: Satia Herfert <satiaherfert@gmx.de>

WilcoNeo committed:
Update modules/ROOT/pages/patterns/shortest-paths.adoc

Co-authored-by: Satia Herfert <satiaherfert@gmx.de>

WilcoNeo committed:
Update modules/ROOT/pages/patterns/shortest-paths.adoc

Co-authored-by: Satia Herfert <satiaherfert@gmx.de>

JPryce-Aklundh commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,90 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+[[planning-StatefulShortestPath-operator]]
+== Planning the StatefulShortestPath-operator
+The `StatefulShortestPath(All)` uses a unidirectional breadth first search algorithm, while the `StatefulShortestPath(Into)` uses a bidirectional breadth first search algorithm.
+See xref:planning-and-tuning/operators/operators-detail.adoc#query-plan-stateful-shortest-path-into[the operator-detail page] for a description of the `StatefulShortestPath(All)` and `StatefulShortestPath(Into)` operators.
```
```suggestion
== Planning shortest paths queries
This section describes the operators used when planning shortest paths queries.
For readers not familiar with Cypher execution plans and operators, it is recommended to first read the section xref:planning-and-tuning/execution-plans.adoc[].

There are two operators used to plan `SHORTEST` queries:

* xref:planning-and-tuning/operators/operators-detail.adoc#query-plan-stateful-shortest-path-all[`StatefulShortestPath(All)`] - uses a unidirectional breadth-first search algorithm to find shortest paths from a previously matched start node to an end node that has not yet been matched.

* xref:planning-and-tuning/operators/operators-detail.adoc#query-plan-stateful-shortest-path-into[`StatefulShortestPath(Into)`] - uses a bidirectional breadth-first search algorithm (BFS) algorithm, where two simultaneous BFS invocations are performed, one from the left boundary node and one from the right boundary node, until a pattern is found.
```

JPryce-Aklundh commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,90 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+
+* When both boundary nodes have been matched in a `MATCH` clause prior to the `MATCH` clause containing the shortest paths query, `StatefulShortestPath(Into)` is planned.
+* Otherwise, if both boundary nodes match to at most one data node, then `StatefulShortestPath(Into)` is planned.
+* Otherwise, `StatefulShortestPath(All)` is planned.
```
```suggestion

`StatefulShortestPath(Into)` is used by the planner when both boundary nodes in the shortest path have been matched in a `MATCH` clause previous to that containing the `SHORTEST` keyword.
It is also used when both boundary nodes match at most one node each.
Otherwise, `StatefulShortestPath(All)` is used.
```

JPryce-Aklundh commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,90 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+* Otherwise, `StatefulShortestPath(All)` is planned.
+
+Sometimes the planner cannot make reliable estimations about how many data nodes a pattern node will match.
+Consider using a xref:constraints/index.adoc#unique-node-property[uniqueness constraint] where applicable to help the planner getting more reliable estimates.
```
```suggestion
```

JPryce-Aklundh commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,90 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+Sometimes the planner cannot make reliable estimations about how many data nodes a pattern node will match.
+Consider using a xref:constraints/index.adoc#unique-node-property[uniqueness constraint] where applicable to help the planner getting more reliable estimates.
+
+As an example, consider the case where the left boundary node matches to 10 data nodes and the right boundary node to 3 data nodes.
+Using `StatefulShortestPath(Into)` requires 30 bidirectional breath first search (BFS) invocations.
+Using `StatefulShortestPath(All)` requires only 3 unidirectional BFS invocations expanding from the right boundary node.
```
```suggestion
For example, the left boundary node in the below query matches 10 nodes, and the right boundary node matches 3 nodes.
Using `StatefulShortestPath(Into)` requires 30 bidirectional breath-first search (BFS) invocations, while `StatefulShortestPath(All)` requires only 3 unidirectional BFS invocations expanding from the right boundary node.
As a result, the query will use `StatefulShortestPath(All)`.

.Query planned with `StatefulShortestPath(All)`
[source,cypher]
----
MATCH
  p = ALL SHORTEST (a:Station)(()-[]-()-[]-()){1,}(b:Station)
RETURN p
----

.Result
[role="queryplan", subs="attributes+"]
----
+----------------------------+----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
| Operator                   | Id | Details                                                                                 | Estimated Rows | Rows | DB Hits | Memory (Bytes) | Page Cache Hits/Misses | Time (ms) | Pipeline      |
+----------------------------+----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
| +ProduceResults            |  0 | p                                                                                       |            383 |  154 |       0 |              0 |                    0/0 |     0.158 |               |
| |                          +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+               |
| +Projection                |  1 | (a) ((anon_12)-[anon_14]-(anon_13)-[anon_11]-())* (b) AS p                              |            383 |  154 |       0 |                |                    0/0 |     2.785 |               |
| |                          +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+               |
| +StatefulShortestPath(All) |  2 | SHORTEST 1 GROUPS (a) ((`anon_5`)-[`anon_6`]-(`anon_7`)-[`anon_8`]-(`anon_9`)){1, } (b) |            383 |  154 |     655 |          19711 |                    1/0 |    70.908 | In Pipeline 1 |
| |                          |    |         expanding from: a                                                               |                |      |         |                |                        |           |               |
| |                          |    |     inlined predicates: b:Station                                                       |                |      |         |                |                        |           |               |
| |                          +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
| +NodeByLabelScan           |  3 | a:Station                                                                               |             10 |    9 |      10 |            376 |                    1/0 |     1.106 | In Pipeline 0 |
+----------------------------+----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
----

```

JPryce-Aklundh commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,90 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+Using `StatefulShortestPath(Into)` requires 30 bidirectional breath first search (BFS) invocations.
+Using `StatefulShortestPath(All)` requires only 3 unidirectional BFS invocations expanding from the right boundary node.
+
+This heuristic decision to favour `StatefulShortestPath(All)` can sometimes lead to worse query performance.
```
```suggestion
However, the heuristic decision to favor `StatefulShortestPath(All)` can lead to worse query performance.
To have the planner choose the `StatefulShortestPath(Into)`, rewrite the query using a xref:subqueries/call-subquery.adoc[`CALL` subquery].
This is because the `CALL` subquery is executed for each incoming row.
Thus, in the below query, both `a` and `b` are matched to exactly one `Station` node respectively which enables the planner to use `StatefulShortestPath(Into)`.

.Query rewritten to use `StatefulShortestPath(Into)`
[source,cypher]
----
MATCH
  (a:Station),
  (b:Station)
CALL {
  WITH a, b
  MATCH
    p = ALL SHORTEST (a)(()-[]-()-[]-()){1,}(b)
  RETURN p
}
RETURN p
----

.Result
role="queryplan", subs="attributes+"]
----
+-----------------------------+----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
| Operator                    | Id | Details                                                                                 | Estimated Rows | Rows | DB Hits | Memory (Bytes) | Page Cache Hits/Misses | Time (ms) | Pipeline      |
+-----------------------------+----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
| +ProduceResults             |  0 | p                                                                                       |            383 |  154 |       0 |              0 |                    0/0 |     0.166 |               |
| |                           +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+               |
| +Projection                 |  1 | (a) ((anon_12)-[anon_14]-(anon_13)-[anon_11]-())* (b) AS p                              |            383 |  154 |       0 |                |                    0/0 |     2.605 |               |
| |                           +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+               |
| +StatefulShortestPath(Into) |  2 | SHORTEST 1 GROUPS (a) ((`anon_5`)-[`anon_6`]-(`anon_7`)-[`anon_8`]-(`anon_9`)){1, } (b) |            383 |  154 |    1517 |          24486 |                    1/0 |    76.973 | In Pipeline 3 |
| |                           |    |         expanding from: a                                                               |                |      |         |                |                        |           |               |
| |                           +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
| +CartesianProduct           |  3 |                                                                                         |            100 |   81 |       0 |           2056 |                    0/0 |     1.352 | In Pipeline 2 |
| |\                          +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
| | +NodeByLabelScan          |  4 | b:Station                                                                               |             10 |    9 |      10 |            392 |                    1/0 |     0.229 | In Pipeline 1 |
| |                           +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
| +NodeByLabelScan            |  5 | a:Station                                                                               |             10 |    9 |      10 |            376 |                    1/0 |     1.224 | In Pipeline 0 |
+-----------------------------+----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
----
```

JPryce-Aklundh commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,90 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+----
+
+
+
+
+Rewriting the query using a `CALL` subquery
+[source,cypher]
+----
+MATCH
+  (a:Station),
+  (b:Station)
+CALL {
+  WITH a, b
+  MATCH
+    p = ALL SHORTEST (a)(()-[]-()-[]-()){1,}(b)
+  RETURN p
+}
+RETURN p
+----
+
+makes use of `StatefulShortestPath(Into)`, because the CALL subquery is executed for each incoming row. Each incoming row matches `a` to exactly one station node and `b` to exactly one station node.
+[role="queryplan", subs="attributes+"]
+----
++-----------------------------+----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
+| Operator                    | Id | Details                                                                                 | Estimated Rows | Rows | DB Hits | Memory (Bytes) | Page Cache Hits/Misses | Time (ms) | Pipeline      |
++-----------------------------+----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
+| +ProduceResults             |  0 | p                                                                                       |            383 |  154 |       0 |              0 |                    0/0 |     0.166 |               |
+| |                           +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+               |
+| +Projection                 |  1 | (a) ((anon_12)-[anon_14]-(anon_13)-[anon_11]-())* (b) AS p                              |            383 |  154 |       0 |                |                    0/0 |     2.605 |               |
+| |                           +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+               |
+| +StatefulShortestPath(Into) |  2 | SHORTEST 1 GROUPS (a) ((`anon_5`)-[`anon_6`]-(`anon_7`)-[`anon_8`]-(`anon_9`)){1, } (b) |            383 |  154 |    1517 |          24486 |                    1/0 |    76.973 | In Pipeline 3 |
+| |                           |    |         expanding from: a                                                               |                |      |         |                |                        |           |               |
+| |                           +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
+| +CartesianProduct           |  3 |                                                                                         |            100 |   81 |       0 |           2056 |                    0/0 |     1.352 | In Pipeline 2 |
+| |\                          +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
+| | +NodeByLabelScan          |  4 | b:Station                                                                               |             10 |    9 |      10 |            392 |                    1/0 |     0.229 | In Pipeline 1 |
+| |                           +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
+| +NodeByLabelScan            |  5 | a:Station                                                                               |             10 |    9 |      10 |            376 |                    1/0 |     1.224 | In Pipeline 0 |
++-----------------------------+----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
+----
```
```suggestion
[TIP]
Sometimes the planner cannot make reliable estimations about how many nodes a pattern node will match.
Consider using a xref:constraints/index.adoc#unique-node-property[uniqueness constraint] where applicable to help the planner get more reliable estimates.
```

JPryce-Aklundh reviewed (commented):
Thanks @wilco - this will be very useful to readers 👍

WilcoNeo committed:
Update modules/ROOT/pages/patterns/shortest-paths.adoc

Co-authored-by: Jens Pryce-Åklundh <112686610+JPryce-Aklundh@users.noreply.github.com>

WilcoNeo committed:
Update modules/ROOT/pages/patterns/shortest-paths.adoc

Co-authored-by: Jens Pryce-Åklundh <112686610+JPryce-Aklundh@users.noreply.github.com>

WilcoNeo committed:
Update modules/ROOT/pages/patterns/shortest-paths.adoc

Co-authored-by: Jens Pryce-Åklundh <112686610+JPryce-Aklundh@users.noreply.github.com>

WilcoNeo committed:
Update modules/ROOT/pages/patterns/shortest-paths.adoc

Co-authored-by: Jens Pryce-Åklundh <112686610+JPryce-Aklundh@users.noreply.github.com>

WilcoNeo committed:
Update modules/ROOT/pages/patterns/shortest-paths.adoc

Co-authored-by: Jens Pryce-Åklundh <112686610+JPryce-Aklundh@users.noreply.github.com>

WilcoNeo committed:
Update modules/ROOT/pages/patterns/shortest-paths.adoc

Co-authored-by: Jens Pryce-Åklundh <112686610+JPryce-Aklundh@users.noreply.github.com>

WilcoNeo committed:
SHORTEST paths planning doc: make example match the explanation

WilcoNeo committed:
SHORTEST paths planning doc: fix explain message of Into

WilcoNeo committed:
SHORTEST paths doc operator details: fix explain message of Into

WilcoNeo committed:
SHORTEST paths doc planning details

JPryce-Aklundh commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,100 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+
+
+== Planning shortest paths queries
+This section describes the operators used when planning shortest paths queries.
```
```suggestion

== Planning shortest paths queries

This section describes the operators used when planning shortest paths queries.
```

dogofbrian replied:
```suggestion
This section describes the operators used when planning shortest path queries.
```

dogofbrian commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,100 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+
+
+For example, the left boundary node in the below query matches 9 nodes, and the right boundary node matches 3 nodes.
+Using `StatefulShortestPath(Into)` requires 27 bidirectional breath-first search (BFS) invocations, while `StatefulShortestPath(All)` requires only 3 unidirectional BFS invocations expanding from the right boundary node.
```
```suggestion
Using `StatefulShortestPath(Into)` requires 27 bidirectional breadth-first search (BFS) invocations, while `StatefulShortestPath(All)` requires only 3 unidirectional BFS invocations expanding from the right boundary node.
```

WilcoNeo committed:
Update modules/ROOT/pages/patterns/shortest-paths.adoc

Co-authored-by: dogofbrian <dogofbrian@users.noreply.github.com>

WilcoNeo committed:
Update modules/ROOT/pages/patterns/shortest-paths.adoc

Co-authored-by: dogofbrian <dogofbrian@users.noreply.github.com>

WilcoNeo committed:
SHORTEST paths doc typo

JPryce-Aklundh commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,100 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
 2+d|Rows: 1
 |===
+
+
+== Planning shortest paths queries
+This section describes the operators used when planning shortest path queries.
```
```suggestion

== Planning shortest paths queries

This section describes the operators used when planning shortest path queries.
```

JPryce-Aklundh commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,100 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+
+
+However, the heuristic decision to favor `StatefulShortestPath(All)` can lead to worse query performance.
+To have the planner choose the `StatefulShortestPath(Into)`, rewrite the query using a xref:subqueries/call-subquery.adoc[`CALL` subquery].
+This is because the `CALL` subquery is executed for each incoming row.
```
```suggestion
To have the planner choose the `StatefulShortestPath(Into)`, rewrite the query using a xref:subqueries/call-subquery.adoc[`CALL` subquery], which is executed for each incoming row.
```

JPryce-Aklundh committed:
Update modules/ROOT/pages/patterns/shortest-paths.adoc

dogofbrian commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,101 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+* xref:planning-and-tuning/operators/operators-detail.adoc#query-plan-stateful-shortest-path-into[`StatefulShortestPath(Into)`] - uses a bidirectional breadth-first search (BFS) algorithm, where two simultaneous BFS invocations are performed, one from the left boundary node and one from the right boundary node.
+
+
+`StatefulShortestPath(Into)` is used by the planner when both boundary nodes in the shortest path have been matched in a `MATCH` clause previous to that containing the `SHORTEST` keyword.
```
The wording makes this seem to be a plan that is decided at runtime, rather than based on cardinality estimates. For example, on the UK railway dataset I get only one node matching each end node pattern, but it is still planned as `StatefulShortestPath(All)`:
```
PROFILE
MATCH (n:Station {name: "Denmark Hill"}), (m:Station WHERE m.name CONTAINS "Vic")
MATCH p = SHORTEST 1 (n)((l)-[e:LINK]-(r) WHERE e.distance < 10){,5}(m)
RETURN p
```

WilcoNeo replied:
Good point. Let me look into this

dogofbrian commented on modules/ROOT/pages/patterns/shortest-paths.adoc:
```diff
@@ -538,3 +538,101 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
+However, the heuristic decision to favor `StatefulShortestPath(All)` can lead to worse query performance.
+To have the planner choose the `StatefulShortestPath(Into)`, rewrite the query using a xref:subqueries/call-subquery.adoc[`CALL` subquery].
+This is because the `CALL` subquery is executed for each incoming row.
+Thus, in the below query, both `a` and `b` are matched to exactly one `Station` node respectively which enables the planner to use `StatefulShortestPath(Into)`.
```
It would be worth clarifying that for each invocation of the subquery `CALL`, the planner knows that `a` and `b` will be bound to one `Station` node each, and thus the planner uses `StatefulShortestPath(Into)`.

WilcoNeo replied:
I have suggested a change in Jens his PR when this one is merged in

dogofbrian replied:
This is addressed in [863](https://github.com/neo4j/docs-cypher/pull/863)

dogofbrian commented on modules/ROOT/pages/planning-and-tuning/operators/operators-detail.adoc:
```diff
@@ -5920,6 +5920,118 @@ Total database accesses: 5, total allocated memory: 1488
+The `StatefulShortestPath(Into)` operator finds shortest paths between two previously matched node variables.
+It uses a bidirectional breadth-first search (BFS) algorithm, which performs two BFS invocations at the same time, one from the left boundary node and one from the right boundary node.
+Once a node is found by both BFS invocations, which indicates that it can be reached from both boundary nodes, the algorithm successfully terminates.
+When both BFS invocations terminate without any overlap in the nodes found, then there exists no valid path between both boundary nodes.
```
Is it both, or just one? I thought that when one of the BFS exhausts all permitted paths from it's starting node, the whole algorithm terminates. If it didn't do that, wouldn't the disconnected cases perform as poorly as when it was implemented with "unidirectional" BFS?

WilcoNeo replied:
I think you are right. Both BFSs run at the same time. But when one terminates without finding any overlap so far, the continuation of the other can't lead to successful path. Do we want to verify this with the run-time team?

sherfert commented:
@dogofbrian valuable comments.
Could you transfer them to the [main PR](https://github.com/neo4j/docs-cypher/pull/863), now that this one got merged?

dogofbrian commented:
Yes, definitely worth verifying with Runtime. If a user is seeking an in-depth understanding of performance of the operators, this point will help them a lot with the disconnected case.
